service/api_logic/recommendation_logic.py

```diff
@@ -0,0 +1,289 @@
+MAX_PROMOTION_COF = 4
+
+
+class RecommendationService:
+    def __init__(self, recommendation_dal, user_dal, news_dal):
+        self._recommendation_dal = recommendation_dal
+        self._user_dal = user_dal
+        self._news_dal = news_dal
+        self.logger = Logger("logger", "all.log").logger
+        self._recommendations_list = []
+
+
+    async def auto_hybrid_recommendation_system(self, top_n=BASE_LIMIT_OF_NEWS_FOR_RECS):
+        users_that_want_recs = self._recommendation_dal.get_all_users()
+        if not users_that_want_recs:
+            raise NoUsersInDBError
+
+        interaction_df = self.__get_interactions(users_that_want_recs)
+        users_interaction_matrix = self.__create_interaction_matrix(interaction_df)
+        for user in users_that_want_recs:
+            try:
+                recommendations_for_user_with_score = await self.__user_based_recommendations(user.user_id, users_interaction_matrix, top_n=top_n)
+                if not recommendations_for_user_with_score or not any(recommendations_for_user_with_score):
+                    raise EmptyRecommendationListForUserError(user.user_id)
+
+                self._recommendations_list.append(recommendations_for_user_with_score)
+            except EmptyRecommendationListForUserError as e:
+                self.logger.warning(f"{e}, so skipping this user")
+                continue
+
+        self.__save_to_db_users_recommendations(self._recommendations_list)
+
+        return self._recommendations_list
+
+
+    async def fast_hybrid_recommendation_system(self, user, top_n=BASE_LIMIT_OF_NEWS_FOR_RECS):
+        interaction_df = self.__get_interactions(user)
+        users_interaction_matrix = self.__create_interaction_matrix(interaction_df)
+        recommendations_for_user_with_score = await self.__user_based_recommendations(user.user_id, users_interaction_matrix, top_n=top_n)
+        if not recommendations_for_user_with_score or not any(recommendations_for_user_with_score):
+            raise EmptyRecommendationListForUserError(user.user_id)
+
+        self._recommendations_list.append(recommendations_for_user_with_score)
+
+        self.__save_to_db_users_recommendations(self._recommendations_list)
+
+        return self._recommendations_list
+
+
+    def __get_interactions(self, users_that_want_recs, time_limit=MAX_LIMIT_FOR_REC_IN_DAYS):
+        time_limit = datetime.now() - timedelta(days=time_limit)
+        interactions = self._recommendation_dal.get_user_interactions(time_limit)
+        interactions_df = pd.DataFrame(interactions, columns=['user_id', 'news_id', 'interaction', 'timestamp'])
+        interactions_grouped = interactions_df.groupby(['user_id', 'news_id'], as_index=False)['interaction'].sum()
+
+        all_news_by_last_const_days = self._news_dal.get_news_ids(time_limit)
+        all_news_by_last_const_days_df = pd.DataFrame(all_news_by_last_const_days, columns=['news_id'])
+
+        if isinstance(users_that_want_recs, list):
```

Comment:
It's our backend and we control everything that's happening here. Can't we just ensure, that this method will always get a list?